aiida/common/test_hashing.py

```diff
@@ -32,72 +54,111 @@ class MakeHashTest(unittest.TestCase):
+            True: '31ad5fa163a0c478d966c7f7568f3248f0c58d294372b2e8f7cb0560d8c8b12f',
+            None: '1729486cc7e56a6383542b1ec73125ccb26093651a5da05e04657ac416a74b8f',
+        }
+
+        for val, digest in test_data.items():
+            with self.subTest(val=val):
+                self.assertEqual(make_hash(val), digest)
 
     def test_unicode_string(self):
         self.assertEqual(
-            make_hash(u'something still in ASCII'), '753df9fb179504652f66dd15877d393a743cbfe31d6bf6a9faed7e37')
-        # current implementation fails with real unicode:
-        # self.assertEqual(make_hash(u'öpis mit Umluut wie ä, ö, ü und emene ß 😊'), '')
-
-    def test_collection_with_builtin_types(self):
-        self.assertEqual(make_hash((1, 2, 3)), 'ccbd4c0c83df304a4d8fa1cb4d556f370d8d09c03b49cd069fd72598')
-        self.assertEqual(make_hash([1, 2, 3]), 'ccbd4c0c83df304a4d8fa1cb4d556f370d8d09c03b49cd069fd72598')
-        self.assertEqual(make_hash([3, 1, 2]), '401c29a918e635ab4b384e7a0ac1d189eb857bfc3e8085fe3aaa17c2')
-        self.assertEqual(make_hash({1, 2, 3}), '95467151c8099d6bef0e52588f0e651d6276fa4cad0866c710447a58')
-        self.assertEqual(make_hash({'a': 'b', 'c': 'd'}), 'ee137be5e9397e30612b6628df24be4e4913c6cc4a6be5bf85509ef3')
-        self.assertEqual(make_hash({'c': 'd', 'a': 'b'}), 'ee137be5e9397e30612b6628df24be4e4913c6cc4a6be5bf85509ef3')
-
-    @unittest.skipIf(six.PY3, "Broken on Python 3")
+            make_hash(u'something still in ASCII'), 'd55e492596cf214d877e165cdc3394f27e82e011838474f5ba5b9824074b9e91')
+
+        self.assertEqual(
+            make_hash(u"öpis mit Umluut wie ä, ö, ü und emene ß"),
+            'c404bf9a62cba3518de5c2bae8c67010aff6e4051cce565fa247a7f1d71f1fc7')
+
+    def test_collection_with_ordered_sets(self):
+        self.assertEqual(make_hash((1, 2, 3)), '27e6598a3bd1c878ca09db22f4d93ebcb1a0e08db27e09c737e7f2e698dcce5b')
+        self.assertEqual(make_hash([1, 2, 3]), '27e6598a3bd1c878ca09db22f4d93ebcb1a0e08db27e09c737e7f2e698dcce5b')
+
+        for perm in itertools.permutations([1, 2, 3]):
+            with self.subTest(orig=[1, 2, 3], perm=perm):
+                self.assertNotEqual(make_hash(perm), make_hash({1, 2, 3}))
+
+    def test_collection_with_unordered_sets(self):
+        self.assertEqual(make_hash({1, 2, 3}), '2e096a4b81ab724c2250835fc2fef7d932a7a86deafe1fb584ff329adcc34e0f')
+        self.assertEqual(make_hash({1, 2, 3}), make_hash({2, 1, 3}))
+
+    def test_collection_with_dicts(self):
+        self.assertEqual(
+            make_hash({
+                'a': 'b',
+                'c': 'd'
+            }), '1890162a70dbbcbef2d0cc40b9bd90463459f2e7e72573628e7f1c003ca25c95')
+
+        # order changes in dictionaries should give the same hashes
+        self.assertEqual(
+            make_hash(collections.OrderedDict([('c', 'd'), ('a', 'b')])),
+            make_hash(collections.OrderedDict([('a', 'b'), ('c', 'd')])))
```

Comment:
Suggestion: for make_hash, when an OrderedDict is passed, check for a kwarg (e.g. 'consider_as_unsorted' or a better name): if True, continue; if False, raise